### PR TITLE
Establish first test: button count

### PR DESCRIPTION
**Pull request by UltimateGeek**
Uses Jest for testing, taking advantage of ESM support which is not quite finished, but usable.

Produces coverage report for the js files that get included.

Test setup initially limited to client.

ESM imports/exports limited to only what's needed to facilitate the initial test. 
Minification of client code updated to strip the imports.

To setup github CI to run the tests:
* [ ] Setup the [Node.js CI workflow](https://github.com/UltimateGeek/virtualtabletop/new/main?filename=.github%2Fworkflows%2Fnode.js.yml&workflow_template=node.js) -- change it to only use version "14.x" before commiting.

### CODE REVIEW

File: client/js/connection.js [hunk 1]
```diff
@@ -2,7 +2,14 @@ let lastTimeout = 1000;
 let connection;
 let messageCallbacks = {};
 
-function startWebSocket() {
+//used by unit tests until jest supports mocking ESM static imports
+export function mockConnection() {
```
Comment (ArnoldSmith86) [inline client/js/connection.js:6]: All these exports/imports still work in the browser? In all relevant browsers?

File: client/js/serverstate.js [hunk 2]
```diff
@@ -10,11 +13,11 @@ let deltaChanged = false;
 let deltaID = 0;
 let batchDepth = 0;
 
-function addWidget(widget) {
+export function addWidget(widget, instance) {
   if(widget.parent && !widgets.has(widget.parent)) {
     if(!deferredChildren[widget.parent])
       deferredChildren[widget.parent] = [];
-    deferredChildren[widget.parent].push(widget);
+    deferredChildren[widget.parent].push(widget); //TODO: save instance
```
Comment (ArnoldSmith86) [inline client/js/serverstate.js:20]: What does "save instance" mean?

File: client/js/widgets/button.js [hunk 3]
```diff
@@ -58,6 +64,7 @@ class Button extends Widget {
       const a = { ...original };
       var problems = [];
 
+      if (this.p('debug')) console.log(`${this.id}: ${JSON.stringify(original)}`);
```
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:67]: Seems to be unrelated to the PR. It's fine, I just wonder what you used it for.

File: client/js/widgets/button.js [hunk 4]
```diff
@@ -82,23 +89,31 @@ class Button extends Widget {
         if(isValidCollection(a.collection))
           for(const w of collections[a.collection])
             w.click();
+        else
```
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:92]: These SHOULD be part of isValidCollection and isValidID. Doesn't the push in there work? I might not have tested it.

File: client/js/widgets/button.js [hunk 5]
```diff
@@ -133,7 +148,10 @@ class Button extends Widget {
         setDefaults(a, { value: 0, mode: 'set' });
         if([ 'set', 'dec', 'inc' ].indexOf(a.mode) == -1)
           problems.push(`Warning: Mode ${a.mode} will be interpreted as add.`);
-        this.w(a.label, label=>label.setText(a.value, a.mode));
+        this.w(a.label, label=> {
+          if (this.p('debug')) console.log(`changing ${a.label} ${a.mode} ${a.value}`)
```
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:152]: I'm guessing this is left over debug code? Or do you react to these with the test framework? I guess I'll find out. :D

File: client/js/widgets/button.js [hunk 6]
```diff
@@ -275,14 +293,17 @@ class Button extends Widget {
       }
 
       if(this.p('debug')) {
-        $('#debugButtonOutput').textContent += '\n\n\nOPERATION: \n' + JSON.stringify(a, null, '  ');
+        var msg = ''
```
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:296]: Not really important but please use const or let instead of var. No need to change it though.

File: server/minify.mjs [hunk 7]
```diff
@@ -64,9 +64,10 @@ export default function minifyRoom() {
         output: '/tmp/out.js'
       });
     }).then(function(min) {
+      const minNoImports = min.replace(/\bimport[^;]*\.\/[^;]*;/g, "")
```
Comment (ArnoldSmith86) [inline server/minify.mjs:67]: Oh, guess that answers one question...

File: tests/client/button-count.test.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
+import { JSDOM } from 'jsdom';
+//import { jest } from '@jest/globals';
+
+import { mockConnection } from '../../client/js/connection.js';
+import { addWidget } from '../../client/js/serverstate.js';
+import { Button } from '../../client/js/widgets/button.js';
+import { Label } from '../../client/js/widgets/label.js';
+
+const loadWindow = async () => {
```
Comment (ArnoldSmith86) [inline tests/client/button-count.test.js:9]: I'm quite new to async/await/Promises but I'm pretty sure this is too complicated. Doesn't `loadWindow = new Promise(...)` work?

File: client/js/connection.js (hunk 1 above)
Comment (UltimateGeek) [reply client/js/connection.js:6]: Not the imports. They are stripped as part of the minify code. Exports work: I've tested Chromium and Safari. 

Where is the list of supported browsers specified?

File: client/js/serverstate.js (hunk 2 above)
Comment (UltimateGeek) [reply client/js/serverstate.js:20]: I added a parameter to pass in an existing instance of the widget to be used. I didn't update this portion of the function to keep track of it when deferring adding the widget. I didn't update this portion of the function because if I did, then I would need to test it. And I didn't have a need to have it support the deferment of a passed in pre-created instance.

So, I left this TODO here to indicate the implementation for fully supporting the new parameter is incomplete. Would you like to see the comment updated to clarify?

File: client/js/widgets/button.js (hunk 3 above)
Comment (UltimateGeek) [reply client/js/widgets/button.js:67]: I found it helpful when debugging my code to have a copy of the original before doing the applyVariables.

File: client/js/widgets/button.js (hunk 5 above)
Comment (UltimateGeek) [reply client/js/widgets/button.js:152]: I used it throughout development of the test, but then disabled usage of debug before committing my changes. For the purpose of achieving higher coverage, we should creating an internal log function that is always called and have it check if debug is on.

File: tests/client/button-count.test.js (hunk 8 above)
Comment (UltimateGeek) [reply tests/client/button-count.test.js:9]: No, but `async function loadWindow() { ... ` does.

File: client/js/widgets/button.js (hunk 4 above)
Comment (UltimateGeek) [reply client/js/widgets/button.js]: You're right. It does do the logging. I missed that. Removed here & in a couple other spots I had added it.

I added a corresponding test for the case where the collection hasn't been created by a prior function.

File: client/js/widgets/button.js (hunk 6 above)
Comment (UltimateGeek) [reply client/js/widgets/button.js]: These nuances with javascript I'm not so familiar with, so please be picky to help me understand subtle importances. :-) I changed it to `let`.

File: tests/client/button-count.test.js (hunk 8 above)
Comment (UltimateGeek) [reply tests/client/button-count.test.js]: I tried getting rid of the `loadWindow()` function, but couldn't get it to work with simpler code than simply declaring it `async`.

File: client/js/serverstate.js (hunk 2 above)
Comment (ArnoldSmith86) [reply client/js/serverstate.js:20]: No, I see what you mean. It's probably never necessary though, right? Why would you feed it hand-created widgets in the wrong order?

File: client/js/widgets/button.js (hunk 6 above)
Comment (ArnoldSmith86) [reply client/js/widgets/button.js]: Yeah, I only learned about those recently as well. `var` has a pretty weird scope. `let` and `const` are valid inside their block which is what I would expect.

**Comment by ArnoldSmith86** [comment]
This is really nice. What do you think about using vtt files for the tests? Should be less work to build the tests, right? We could even use the tutorial files for it.

And atm this would pretty much only allow to test buttons right because you can't simulate mouse movements or something similar? But I imagine that's a lot more work...

File: client/js/connection.js (hunk 1 above)
Comment (ArnoldSmith86) [reply client/js/connection.js:6]: There's no list but I generally just check [caniuse.com](https://www.caniuse.com/?search=export). It looks like it's not supported on the current Android Browser which should be a deal breaker (unless it simply doesn't understand it but everything still works).

**Comment by UltimateGeek** [comment]
> This is really nice. What do you think about using vtt files for the tests? Should be less work to build the tests, right? We could even use the tutorial files for it.

What you suggest about using vtt files is what I would consider an integration test. Testing the tutorials definitely has the benefit of alerting us if there is regression that needs to be fixed.

However, that's a bit more work. Have to use something like Selenium to automate the web browser and user inputs.

> And atm this would pretty much only allow to test buttons right because you can't simulate mouse movements or something similar? But I imagine that's a lot more work...

I think it's possible to simulate mouse movements. But yes, something like Selenium is yet another tool to have to learn.

Here's a tutorial for getting it setup, integrated with jest: https://www.lambdatest.com/blog/jest-tutorial-for-selenium-javascript-testing-with-examples/

File: client/js/connection.js (hunk 1 above)
Comment (UltimateGeek) [reply client/js/connection.js:6]: I don't have an android device. Per these [chrome-devtools "emulate & test other browsers"](https://developers.google.com/web/tools/chrome-devtools/device-mode/testing-other-browsers) instructions, it talks about downloading an Android emulator by way of the Android SDK. Which I guess that means, I can't just use F12 in Chromium to put the browser into an Android webkit emulation mode.

Does that sound right to you? If so, I'll take the time to install and try it out.

File: client/js/serverstate.js (hunk 2 above)
Comment (UltimateGeek) [reply client/js/serverstate.js:20]: Yes, true. we can probably just make sure unit tests create & add items in the needed order.

So, what do you want me to do about the TODO comment?

File: client/js/connection.js (hunk 1 above)
Comment (ArnoldSmith86) [reply client/js/connection.js:6]: I think we'll have enough Android devices in the community. I will test with mine when I have the time.

File: client/js/serverstate.js (hunk 2 above)
Comment (ArnoldSmith86) [reply client/js/serverstate.js:20]: I would remove it but doesn't really matter.

File: client/js/widgets/button.js (hunk 5 above)
Comment (ArnoldSmith86) [reply client/js/widgets/button.js:146]: I don't really care. I would just like the console to be clean if no button is being debugged.

**Comment by ArnoldSmith86** [comment]
Wouldn't those "higher level" tests make these obsolete though? Or would you do both? That sounds like waaay too much work.

**Comment by ArnoldSmith86** [comment]
You can play around with it here: http://212.47.248.129:8158/ukl9 (or any other room on that server)

**Comment by UltimateGeek** [comment]
@ArnoldSmith86 , differences of the two types of tests:

## unit tests 
1. are easier to write and maintain
2. test specific details and can cover many variations of inputs into functions
3. help ensure you have coverage on as many branches & lines of code as possible
4. run very quickly
5. should not involve any interactions with other processes or network services

## integration tests
1. are more difficult to write and maintain
2. are intended to ensure the integration of things together continue to work
3. can include user acceptance tests
4. typically take a lot longer to run than unit tests
5. can involve interactions with multiple processes and services

Integration tests tend to be more fragile and susceptible to timing changes.

File: client/js/serverstate.js (hunk 2 above)
Comment (UltimateGeek) [reply client/js/serverstate.js:20]:
```suggestion
    deferredChildren[widget.parent].push(widget);
```

**Comment by ArnoldSmith86** [comment]
On the server I get this when starting node:

```
node:internal/process/promises:225
          triggerUncaughtException(err, true /* fromPromise */);
          ^
JS_Parse_Error [SyntaxError]: playerName redeclared
    at JS_Parse_Error.get (eval at <anonymous> (/home/safeuser/staging-PR-10/node_modules/uglify-es/tools/node.js:21:1), <anonymous>:75:23)
    at formatError (node:internal/util/inspect:1179:38)
    at formatRaw (node:internal/util/inspect:949:14)
    at formatValue (node:internal/util/inspect:803:10)
    at inspect (node:internal/util/inspect:336:10)
    at afterInspector (node:internal/errors:721:14)
    at processPromiseRejections (node:internal/process/promises:225:11)
    at processTicksAndRejections (node:internal/process/task_queues:94:32) {
  filename: '0',
  line: 2429,
  col: 4,
  pos: 70605
}
```

Any ideas?

**Comment by UltimateGeek** [comment]
server non-debug startup fixed in 8ef440523

**Comment by ArnoldSmith86** [comment]
I just tested this with Firefox and Chromium on Linux and Android Browser, Privacy Browser and Fennec on Android 10. Seems to be working on all of those.

Did we still have something else to fix/change/test?

**Comment by UltimateGeek** [comment]
Nothing else to test. Should be ready to merge

**Review by ArnoldSmith86 — APPROVED** [review]
